nemo/utils/memory_profile_analyzer.py

```diff
@@ -0,0 +1,473 @@
+import pickle
+import os
+import csv
+import pandas as pd
+import numpy as np
+from scipy.signal import find_peaks, peak_prominences
+from nemo.utils import logging
+
+# __all__ = ['peak_memory_analysis_weight', 'peak_memory_analysis_activation', 'peak_memory_analysis_oom']
+_all__ = ['peak_memory_analysis']
+
+GB_SIZE = 1024*1024*1024
+MB_SIZE = 1024*1024
+KB_SIZE = 1024
+
+def load_pickle_file(filename):
+    with open(filename, "rb") as f:
+        data = pickle.load(f)
+    return data
+
+def make_hashable(obj):
+    """
+    Helper function to make an object hashable. 
+    """
+    if isinstance(obj, dict):
+        return tuple(sorted((k, make_hashable(v)) for k, v in obj.items()))
+    if isinstance(obj, list):
+        return tuple(make_hashable(e) for e in obj)
+    return obj
+
+def prune_frames(frames):
+    """
+    The stack trace is too long, and include many not-so-useful frame. We prune those frames with '??' in the filename. 
+    """
+    return [frame for frame in frames if '??' not in frame['filename']]
+
+def get_printed_frames(alive_memory):
+    """
+    Get the printed frames for the alive_memory. 
+    prune_frames to get rid of '??' in the filename. 
+    Add `\n` in between each frame for the readability. 
+    """
+    # Prune Frames
+    after_prune_frames = [prune_frames(x[3]) for x in alive_memory]
+    # add `\n` to each frame. Connect them together.
+    printed_frames = ['\n'.join([str(frame) for frame in frames]) for frames in after_prune_frames]
+    return printed_frames
+
+
+def read_tp(timepoint):
+    time_us = timepoint['time_us']
+    addr = timepoint['addr']
+    action = timepoint['action']
+    size = timepoint['size'] / GB_SIZE
+    frames = timepoint['frames']
+    stream = timepoint['stream']
+    return (time_us, addr, action, size, frames, stream)
+
+
+def first_py_frame(alloc_frames):
+    """
+    Return the frame when it first shows a `.py` file.
+    """
+    # target_string = 'layernorm_linear.py'
+    for frame in alloc_frames:
+        if '.py' in frame['filename']:
+            return frame
+    return None
+
+def alloc_memory_timeline(trace):
```

Comment:
if that finds the maximum/min alloc memory and the corresponding timestamps, it would be helpful if that was reflected in the name.